From 6bfe2943eb5194dfd1829670d4c45e7c97a7aedf Mon Sep 17 00:00:00 2001
From: Matt Fleming <matt.fleming@intel.com>
Date: Fri, 9 May 2014 14:53:09 +0100
Subject: [PATCH] add done command

---
 psplash.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/psplash.c b/psplash.c
index 09cf0d0..b7eeafe 100644
--- a/psplash.c
+++ b/psplash.c
@@ -121,6 +121,17 @@ parse_command (PSplashFB *fb, char *string, int length)
     {
       return 1;
     }
+  else if (!strcmp(command,"DONE"))
+    {
+      int x, y;
+
+      x = (fb->width - BAR_IMG_WIDTH)/2;
+      y= fb->height - (fb->height/6);
+      psplash_fb_draw_rect (fb, x, y, x+BAR_IMG_WIDTH, y+BAR_IMG_HEIGHT,
+        PSPLASH_BACKGROUND_COLOR);
+
+
+    }
 
   return 0;
 }
@@ -277,6 +288,7 @@ main (int argc, char** argv)
 			 POKY_IMG_RLE_PIXEL_DATA);
 
   /* Draw progress bar border */
+#if 0
   psplash_fb_draw_image (fb, 
 			 (fb->width  - BAR_IMG_WIDTH)/2, 
 			 fb->height - (fb->height/6), 
@@ -284,6 +296,7 @@ main (int argc, char** argv)
 			 BAR_IMG_HEIGHT,
 			 BAR_IMG_BYTES_PER_PIXEL,
 			 BAR_IMG_RLE_PIXEL_DATA);
+#endif
 
   psplash_draw_progress (fb, 0);
 
-- 
1.9.0

