From ed6c9112ee48c7cc8b1f6f737c6c7b1ce38b11a6 Mon Sep 17 00:00:00 2001
From: Matt Fleming <matt.fleming@intel.com>
Date: Fri, 9 May 2014 12:01:37 +0100
Subject: [PATCH] Add luv color background

Signed-off-by: Matt Fleming <matt.fleming@intel.com>
---
 psplash-colors.h | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/psplash-colors.h b/psplash-colors.h
index d701089..ea3b529 100644
--- a/psplash-colors.h
+++ b/psplash-colors.h
@@ -20,15 +20,15 @@
 #define _HAVE_PSPLASH_COLORS_H
 
 /* This is the overall background color */
-#define PSPLASH_BACKGROUND_COLOR 0xec,0xec,0xe1
+#define PSPLASH_BACKGROUND_COLOR 0x17,0x71,0xb7
 
 /* This is the color of any text output */
-#define PSPLASH_TEXT_COLOR 0x6d,0x6d,0x70
+#define PSPLASH_TEXT_COLOR 0xff,0xff,0xff
 
 /* This is the color of the progress bar indicator */
-#define PSPLASH_BAR_COLOR 0x6d,0x6d,0x70
+#define PSPLASH_BAR_COLOR 0xf2,0x73,0x51
 
 /* This is the color of the progress bar background */
-#define PSPLASH_BAR_BACKGROUND_COLOR 0xec,0xec,0xe1
+#define PSPLASH_BAR_BACKGROUND_COLOR 0xff,0xff,0xff
 
 #endif
-- 
1.9.0

